src/core/init/stream_events_emitter/index.ts

```diff
@@ -0,0 +1,164 @@
+ * @returns {Observable}
+ */
+function streamEventsEmitter(manifest: Manifest,
+                             mediaElement: HTMLMediaElement,
+                             clock$: Observable<IInitClockTick>
+): Observable<IStreamEvent> {
+  const eventsBeingPlayed =
+    new WeakMap<IStreamEventPayload|IUnfiniteStreamEventPayload, true>();
+  const scheduledEvents$ = fromEvent(manifest, "manifestUpdate").pipe(
+    startWith(null),
+    scan((oldScheduleEvents) => {
+      return refreshScheduledEventsList(oldScheduleEvents, manifest);
+    }, [] as Array<IStreamEventPayload|IUnfiniteStreamEventPayload>)
+  );
+
+  return scheduledEvents$.pipe(switchMap(poll$));
+
+  function poll$(
+    newScheduleEvents: Array<IStreamEventPayload|IUnfiniteStreamEventPayload>
+  ) {
+    if (newScheduleEvents.length === 0) {
+      return EMPTY;
+    }
+    return observableCombineLatest([
+      interval(STREAM_EVENT_EMITTER_POLL_INTERVAL),
+      clock$,
+    ]).pipe(
+        map(([_, clockTick]) => {
+          const { seeking } = clockTick;
+          return { isSeeking: seeking,
+                   currentTime: mediaElement.currentTime };
+        }),
+        pairwise(),
+        mergeMap(([ oldTick, newTick ]) => {
+          const { isSeeking: wasSeeking, currentTime: previousTime } = oldTick;
+          const { isSeeking, currentTime } = newTick;
+          const eventsToSend: IStreamEvent[] = [];
+          for (let i = newScheduleEvents.length - 1; i >= 0; i--) {
+            const event = newScheduleEvents[i];
+            const start = event.start;
+            const end = isFiniteStreamEvent(event) ? event.end :
+                                                     undefined;
+            const isBeingPlayed = eventsBeingPlayed.has(event);
+            if (isBeingPlayed) {
+              if (start > currentTime ||
+                  (end !== undefined && currentTime >= end)
+              ) {
+                if (isFiniteStreamEvent(event) &&
+                    event.publicEvent.onExit !== undefined &&
+                    typeof event.publicEvent.onExit === "function") {
+                  event.publicEvent.onExit();
+                }
+                eventsBeingPlayed.delete(event);
+              }
+            } else if (start <= currentTime &&
+                       end !== undefined &&
+                       currentTime < end) {
+              eventsToSend.push({ type: "stream-event",
+                                  value: event.publicEvent });
+              eventsBeingPlayed.set(event, true);
+            } else if (isEventIncludedBetweenTimes(event, previousTime, currentTime)) {
+              if (isSeeking && !wasSeeking) {
+                eventsToSend.push({ type: "stream-event-skip",
+                                    value: event.publicEvent });
+              } else {
+                eventsToSend.push({ type: "stream-event",
+                                    value: event.publicEvent });
```

Comment:
Why don't we call `onExit` here?

Reply:
Maybe we should indeed call it if the end is defined.

Reply:
By the way, maybe we should wait for the next poll to call it, as the user may not have declared the onExit callback, as the event has not been sent yet.

Reply:
Oh you're right!

But maybe we can do two birds one stone here.
 
 1. We know that onExit is problematic in general because
 2. We cannot call onExit until the event has been emitted

Maybe we can do each onExit call in a deferred observable triggered after observableOf(...eventsToSend) is triggered.